File: datasets/C++/Claude/DataStructures/medium5.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <climits>
#include <sstream>
using namespace std;

struct TreeNode {
    int data;
    TreeNode* left;
    TreeNode* right;
    
    TreeNode(int val) : data(val), left(nullptr), right(nullptr) {}
};

bool validate(TreeNode* node, long minVal, long maxVal) {
    if (!node) return true;
    
    if (node->data <= minVal || node->data >= maxVal) {
        return false;
    }
    
    return validate(node->left, minVal, node->data) && 
           validate(node->right, node->data, maxVal);
}

bool checkBST(TreeNode* root) {
    return validate(root, LONG_MIN, LONG_MAX);
}

TreeNode* buildTree() {
    cout << "Enter the tree nodes level by level (use -1 for null nodes):" << endl;
    cout << "Example: 3 5 2 1 4 6 -1 -1 -1 -1 -1 -1 -1" << endl;
    
    vector<int> values;
    string line;
    getline(cin, line);
    
    stringstream ss(line);
    int val;
    while (ss >> val) {
        values.push_back(val);
    }
    
    if (values.empty() || values[0] == -1) {
        return nullptr;
    }
    
    TreeNode* root = new TreeNode(values[0]);
    queue<TreeNode*> q;
    q.push(root);
    
    int i = 1;
    while (!q.empty() && i < values.size()) {
        TreeNode* node = q.front();
        q.pop();
        
        // Left child
        if (i < values.size() && values[i] != -1) {
            node->left = new TreeNode(values[i]);
            q.push(node->left);
        }
        i++;
        
        // Right child
        if (i < values.size() && values[i] != -1) {
            node->right = new TreeNode(values[i]);
            q.push(node->right);
        }
        i++;
    }
    
    return root;
}

void printTree(TreeNode* root, int level = 0, string prefix = "Root: ") {
    if (root != nullptr) {
        cout << string(level * 4, ' ') << prefix << root->data << endl;
        if (root->left != nullptr || root->right != nullptr) {
            printTree(root->left, level + 1, "L--- ");
            printTree(root->right, level + 1, "R--- ");
        }
    }
}

int main() {
    TreeNode* root = buildTree();
    
    if (root) {
        cout << "\nBuilt tree:" << endl;
        printTree(root);
        
        bool result = checkBST(root);
        cout << "\nIs BST: " << (result ? "Yes" : "No") << endl;
    } else {
        cout << "Empty tree - considered valid BST: Yes" << endl;
    }
    
    return 0;
}

File: datasets/Python/Claude/DataStructures/medium5.py
from collections import deque

class TreeNode:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def checkBST(root):
    def validate(node, min_val, max_val):
        if not node:
            return True
        
        if node.data <= min_val or node.data >= max_val:
            return False
        
        return (validate(node.left, min_val, node.data) and 
                validate(node.right, node.data, max_val))
    
    return validate(root, float('-inf'), float('inf'))

def build_tree_from_input():
    print("Enter the tree nodes level by level (use -1 for null nodes):")
    print("Example: 3 5 2 1 4 6 -1 -1 -1 -1 -1 -1 -1")
    
    values = list(map(int, input().split()))
    
    if not values or values[0] == -1:
        return None
    
    root = TreeNode(values[0])
    queue = deque([root])
    i = 1
    
    while queue and i < len(values):
        node = queue.popleft()
        
        # Left child
        if i < len(values) and values[i] != -1:
            node.left = TreeNode(values[i])
            queue.append(node.left)
        i += 1
        
        # Right child
        if i < len(values) and values[i] != -1:
            node.right = TreeNode(values[i])
            queue.append(node.right)
        i += 1
    
    return root

def print_tree(root, level=0, prefix="Root: "):
    if root is not None:
        print(" " * (level * 4) + prefix + str(root.data))
        if root.left is not None or root.right is not None:
            print_tree(root.left, level + 1, "L--- ")
            print_tree(root.right, level + 1, "R--- ")

# Main execution
if __name__ == "__main__":
    root = build_tree_from_input()
    
    if root:
        print("\nBuilt tree:")
        print_tree(root)
        
        result = checkBST(root)
        print(f"\nIs BST: {'Yes' if result else 'No'}")
    else:
        print("Empty tree - considered valid BST: Yes")